### DISCUSSION TITLE
How to reliably get type information programically without polling and skipping `{unknown}` response?

### DISCUSSION THREAD

**Discussion by oxalica** [rust-lang/rust-analyzer#17490] [Q&A] [answered]
I'm running rust-analyzer in headless neovim in automation tests and expects `hover` request to return correct types. But it seems sometimes `hover` responses `{unknown}` even after completion of `rustAnalyzer/Indexing`. Adding an infinite `sleep(1) && hover()` works but looks sub-optimal to me.

Is there any notification we can reliable wait for, to guarantee the type information can be retrieved at that point? Or is there any request other than `hover()` which can wait for the type to be calculated before responding the result eagerly as `{unknown}`?

Context: https://github.com/NixOS/nixpkgs/pull/320489#issuecomment-2186555930
The test code: https://github.com/NixOS/nixpkgs/blob/e33ec60b884a7dd7d1fbdcf4cf9a86e1dacf6458/pkgs/development/tools/rust/rust-analyzer/test-neovim-lsp.nix#L13

**Comment by Veykril** [comment, marked as answer]
https://github.com/rust-lang/rust-analyzer/blob/master/docs/dev/lsp-extensions.md#server-status should help I believe, once that returns with `quiescent = true` you should be good to go I think

**Reply by oxalica** [reply]
Thanks, that works for me!
